### PR TITLE
Add weighted sample

### PR DESCRIPTION
**Pull request by DesmonDay**
### PR types
New features

### PR changes
APIs

### Description
Add geometric api: weighted_sample_neighbors.

### CODE REVIEW

File: paddle/fluid/framework/fleet/heter_ps/cudf/block_radix_topk.cuh [hunk 1]
```diff
@@ -0,0 +1,352 @@
+/*
```
Comment (DesmonDay) [inline paddle/fluid/framework/fleet/heter_ps/cudf/block_radix_topk.cuh:1]: 因为这个文件在后续的PGLBox项目中也会用到，所以放置在这个位置。

File: paddle/fluid/framework/fleet/heter_ps/cudf/random.cuh [hunk 2]
```diff
@@ -0,0 +1,80 @@
+/*
```
Comment (DesmonDay) [inline paddle/fluid/framework/fleet/heter_ps/cudf/random.cuh:1]: 同上。

File: python/paddle/geometric/sampling/neighbors.py [hunk 3]
```diff
@@ -170,3 +170,94 @@ def sample_neighbors(
...
     return out_neighbors, out_count
+
+
+def weighted_sample_neighbors(
+    row,
+    colptr,
+    edge_weight,
+    input_nodes,
+    sample_size=-1,
+    eids=None,
```
Comment (YuanRisheng) [inline python/paddle/geometric/sampling/neighbors.py:181]: 这里的sample_size和eids能否调换一下位置，为了和yaml中配置的参数列表顺序一致

File: paddle/phi/kernels/gpu/weighted_sample_neighbors_kernel.cu [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+#else
+#include <cuda_runtime.h>
+#include <curand_kernel.h>
+#include "cub/cub.cuh"
+#endif
+
+#include "math.h"  // NOLINT
+#include "paddle/fluid/framework/fleet/heter_ps/cudf/block_radix_topk.cuh"
+#include "paddle/fluid/framework/fleet/heter_ps/cudf/random.cuh"
+#include "paddle/fluid/memory/memory.h"
```
Comment (YuanRisheng) [inline paddle/phi/kernels/gpu/weighted_sample_neighbors_kernel.cu]: 当前我们在做fluid头文件解耦工作，不能引入fluid下头文件到phi算子库，如果这些头文件是必须的且无法移动到phi算子库下，那么这个kernel文件实现需要放到fluid目录下

File: paddle/phi/infermeta/multiary.h [hunk 5]
```diff
@@ -304,6 +304,17 @@ void GraphSampleNeighborsInferMeta(const MetaTensor& row,
                                    MetaTensor* out_count,
                                    MetaTensor* out_eids);
 
+void WeightedSampleNeighborsInferMeta(const MetaTensor& row,
```
Comment (zyfncg) [inline paddle/phi/infermeta/multiary.h]: 这个函数按字典序放置

File: paddle/phi/kernels/cpu/weighted_sample_neighbors_kernel.cc [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+#include <cmath>
+#include <queue>
+#include <vector>
+
+#include "paddle/phi/backends/cpu/cpu_context.h"
+#include "paddle/phi/core/kernel_registry.h"
+
+template <typename T>
+struct GraphWeightedNode {
```
Comment (zyfncg) [inline paddle/phi/kernels/cpu/weighted_sample_neighbors_kernel.cc:27]: 这个结构体也放到phi的namespace下吧

File: python/paddle/fluid/tests/unittests/test_weighted_sample_neighbors.py [hunk 7]
```diff
@@ -0,0 +1,217 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zyfncg) [inline python/paddle/fluid/tests/unittests/test_weighted_sample_neighbors.py]: 2022 -> 2023

File: paddle/phi/infermeta/multiary.h (hunk 5 above)
Comment (DesmonDay) [reply paddle/phi/infermeta/multiary.h]: done

File: paddle/phi/kernels/gpu/weighted_sample_neighbors_kernel.cu (hunk 4 above)
Comment (DesmonDay) [reply paddle/phi/kernels/gpu/weighted_sample_neighbors_kernel.cu]: 好的，我把这两个文件挪到了paddle/phi/kernels/funcs下面。

File: python/paddle/geometric/sampling/neighbors.py (hunk 3 above)
Comment (DesmonDay) [reply python/paddle/geometric/sampling/neighbors.py:181]: 个人觉得不进行调换比较好。一方面是因为sample_size这个参数比eids重要，想给用户这个感知；另一个是因为已经有个叫sample_neighbors的API，采用的是类似的参数顺序。

File: paddle/phi/kernels/cpu/weighted_sample_neighbors_kernel.cc (hunk 6 above)
Comment (DesmonDay) [reply paddle/phi/kernels/cpu/weighted_sample_neighbors_kernel.cc:27]: done

File: python/paddle/fluid/tests/unittests/test_weighted_sample_neighbors.py (hunk 7 above)
Comment (DesmonDay) [reply python/paddle/fluid/tests/unittests/test_weighted_sample_neighbors.py]: done

**Review by YuanRisheng — APPROVED** [review]

File: python/paddle/geometric/sampling/neighbors.py [hunk 8]
```diff
@@ -170,3 +170,151 @@ def sample_neighbors(
...
+
+    This API is mainly used in Graph Learning domain, and the main purpose is to
+    provide high performance of graph weighted-sampling method. For example, we get the
+    CSC(Compressed Sparse Column) format of the input graph edges as `row` and
+    `colptr`, so as to convert graph data into a suitable format for sampling, and the
+    input `edge_weight` should also match the CSC format. Besides, `input_nodes` means
+    the nodes we need to sample neighbors, and `sample_sizes` means the number of neighbors
+    and number of layers we want to sample. This API will finally return the weighted sampled
+    neighbors, and the probability of being selected as a neighbor is related to its weight,
+    with higher weight and higher probability.
```
Comment (jzhang533) [inline python/paddle/geometric/sampling/neighbors.py:196]: Randomly sample neighbor nodes based on edge weights.

rewriting of this text by ChatGPT:

This API is primarily used in the domain of Graph Learning, with the main purpose of providing high-performance graph weighted-sampling methods. For example, it takes the input graph edges in Compressed Sparse Column (CSC) format as `row` and `colptr`, which allows for converting the graph data into a suitable format for sampling. The `edge_weight` input should also match the CSC format. Additionally, the `input_nodes` parameter specifies the nodes from which we need to sample neighbors, while `sample_sizes` determines the number of neighbors and layers to sample. Finally, this API returns the weighted sampled neighbors, where the probability of being selected as a neighbor is influenced by its weight, with higher weight resulting in higher probability.

**Review by jzhang533 — APPROVED** [review]
LGTM
我试着用ChatGPT修改了一下英文文档。

**Review by zyfncg — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**